numpy/core/src/multiarray/ctors.c

```diff
@@ -1929,7 +1929,149 @@ PyArray_FromAny(PyObject *op, PyArray_Descr *newtype, int min_depth,
+                            "object of too small depth for desired array");
+            ret = NULL;
+        }
+        else if (max_depth != 0 && ndim > max_depth) {
+            Py_DECREF(dtype);
+            Py_XDECREF(newtype);
+            PyErr_SetString(PyExc_ValueError,
+                            "object too deep for desired array");
+            ret = NULL;
+        }
+        else if (ndim == 0 && PyArray_IsScalar(op, Generic)) {
+            ret = (PyArrayObject *)PyArray_FromScalar(op, newtype);
+            Py_DECREF(dtype);
+        }
+        else {
+            if (newtype == NULL) {
+                newtype = dtype;
+            }
+            else {
+                /*
+                 * TODO: would be nice to do this too, but it's
+                 *       a behavior change.  It's also a bit tricky
+                 *       for downcasting to small integer and float
+                 *       types, and might be better to modify
+                 *       PyArray_AssignFromSequence and descr->f->setitem
+                 *       to have a 'casting' parameter and
+                 *       to check each value with scalar rules like
+                 *       in PyArray_MinScalarType.
+                 */
+                /*
+                if (!(flags&NPY_ARRAY_FORCECAST) && ndim > 0 &&
+                        !PyArray_CanCastTo(dtype, newtype)) {
+                    Py_DECREF(dtype);
+                    Py_XDECREF(newtype);
+                    PyErr_SetString(PyExc_TypeError,
+                                "object cannot be safely cast to array "
+                                "of required type");
+                    return NULL;
+                }
+                */
+                Py_DECREF(dtype);
+            }
+
+            /* Create an array and copy the data */
+            ret = (PyArrayObject *)PyArray_NewFromDescr(&PyArray_Type, newtype,
+                                         ndim, dims,
+                                         NULL, NULL,
+                                         flags&NPY_ARRAY_F_CONTIGUOUS, NULL);
+            if (ret == NULL) {
+                return NULL;
+            }
+
+            if (ndim > 0) {
+                if (PyArray_AssignFromSequence(ret, op) < 0) {
+                    Py_DECREF(ret);
+                    ret = NULL;
+                }
+            }
+            else {
+                if (PyArray_SETITEM(ret, PyArray_DATA(ret), op) < 0) {
+                    Py_DECREF(ret);
+                    ret = NULL;
+                }
+            }
+        }
+    }
+    else {
+        if (min_depth != 0 && PyArray_NDIM(arr) < min_depth) {
+            PyErr_SetString(PyExc_ValueError,
+                            "object of too small depth for desired array");
+            Py_DECREF(arr);
+            Py_XDECREF(newtype);
+            ret = NULL;
+        }
+        else if (max_depth != 0 && PyArray_NDIM(arr) > max_depth) {
```

Comment:
Zero is a valid depth - we probably need a different value to mean "not specified", perhaps -1.

Reply:
@eric-wieser Do you mean the `max_depth` or the `ndmax` argument? I duplicated the function `PyArray_FromAny` to add a new parameter, so zero-`max_depth` was already there. Anyway I'm gonna refactor the code to avoid duplication

Reply:
What I mean is, I'd like to see a test for `np.array([1], ndmax=0).ndim == 0`

Reply:
@eric-wieser I made it so and added the test